drivers/soft_spi/soft_spi.c

```diff
@@ -0,0 +1,207 @@
+        gpio_clear(soft_spi_config[bus].mosi_pin);
+    }
+}
+
+int spi_init_cs(spi_t bus, spi_cs_t cs)
+{
+    DEBUG("Soft SPI init CS\n");
+    if (!spi_bus_is_valid(bus)) {
+        DEBUG("Soft SPI bus not valid\n");
+        return SPI_NODEV;
+    }
+
+    if (cs == GPIO_UNDEF) {
+        if (soft_spi_config[bus].cs_pin == GPIO_UNDEF) {
+            DEBUG("Soft SPI no valid CS pin\n");
+            return SPI_NOCS;
+        }
+        DEBUG("Soft SPI set default CS line\n");
+        gpio_init(soft_spi_config[bus].cs_pin, GPIO_OUT);
+        gpio_set(soft_spi_config[bus].cs_pin);
+    }
+    else {
+        DEBUG("Soft SPI set user CS line\n");
+        gpio_init(cs, GPIO_OUT);
+        gpio_set(cs);
+    }
+
+    return SPI_OK;
+}
+
+int spi_acquire(spi_t bus, spi_cs_t cs, spi_mode_t mode, spi_clk_t clk)
+{
+    /* Right now we don't care about speed */
+    (void)clk;
+
+    assert(spi_bus_is_valid(bus));
+
+    /* lock bus */
+    mutex_lock(&locks[bus]);
+
+    if (mode < SPI_MODE_0 || SPI_MODE_3 < mode) {
+        return SPI_NOMODE;
+    }
+    soft_spi_config[bus].spi_mode = mode;
+    switch (mode) {
+        case SPI_MODE_0:
+        case SPI_MODE_1:
+            /* CPOL=0 */
+            gpio_clear(soft_spi_config[bus].clk_pin);
+            break;
+        case SPI_MODE_2:
+        case SPI_MODE_3:
+            /* CPOL=1 */
+            gpio_set(soft_spi_config[bus].clk_pin);
+            break;
+    }
+    return SPI_OK;
+}
+
+void spi_release(spi_t bus)
+{
+    assert(spi_bus_is_valid(bus));
+    mutex_unlock(&locks[bus]);
+}
+
+uint8_t _transfer_one_byte(spi_t bus, spi_cs_t cs, bool cont, uint8_t out)
+{
+    assert(spi_bus_is_valid(bus));
+
+    int8_t bit = 0, i = 0;
+    if (SPI_MODE_1 == soft_spi_config[bus].spi_mode ||
+        SPI_MODE_3 == soft_spi_config[bus].spi_mode) {
+        /* CPHA = 1*/
+        gpio_toggle(soft_spi_config[bus].clk_pin);
+    }
+
+    bit = (out & (1 << 7)) >> 7;
+    gpio_write(soft_spi_config[bus].mosi_pin, bit);
+    for (i = 6; i >= 0; i--) {
+        gpio_toggle(soft_spi_config[bus].clk_pin);
+        gpio_toggle(soft_spi_config[bus].clk_pin);
+        bit = (out & (1 << i)) >> i;
+        gpio_write(soft_spi_config[bus].mosi_pin, bit);
+    }
+    gpio_toggle(soft_spi_config[bus].clk_pin);
+
+    if (SPI_MODE_0 == soft_spi_config[bus].spi_mode ||
+        SPI_MODE_2 == soft_spi_config[bus].spi_mode) {
+        /* CPHASE = 1 */
+        gpio_toggle(soft_spi_config[bus].clk_pin);
+    }
+
+    return out;
+}
+
+void spi_transfer_bytes(spi_t bus, spi_cs_t cs, bool cont,
+                        const void *out, void *in, size_t len)
+{
+    DEBUG("Soft SPI spi_transfer_bytes\n");
+    uint8_t retval = 0;
+
+    /* active the given chip select line */
+    if ((cs != GPIO_UNDEF) && (cs != SPI_CS_UNDEF)) {
+        gpio_clear((gpio_t)cs);
+    }
+    else{
+        gpio_clear(soft_spi_config[bus].cs_pin);
+    }
+
+    for (size_t i = 0; i < len; i++) {
+        if (NULL != out) {
```

Comment:
I think it would make sense to change this block to the following:
```c
uint8_t tmp = (out) ? out[i] : 0;
retval = _transfer_one_byte(bus, cs, true, &tmp);
```
This would also make the inlining more efficient I think.

Reply:
✓